src/lib/cache/SubscriptionCache.cpp

```diff
@@ -0,0 +1,1000 @@
+*
+* Subscription::attributeMatch - 
+*/
+bool Subscription::attributeMatch(const std::string& attributeName)
+{
+  LM_M(("KZ: attributeMatch: %d notifyConditions", notifyConditionVector.size()));
+
+  for (unsigned int ncvIx = 0; ncvIx < notifyConditionVector.size(); ++ncvIx)
+  {
+    NotifyCondition* ncP = notifyConditionVector[ncvIx];
+
+    if (strcasecmp(ncP->type.c_str(), "ONCHANGE") != 0)
+    {
+      LM_M(("KZ: attributeMatch: not ONCHANGE"));
+      continue;
+    }
+
+    LM_M(("KZ: attributeMatch: %d condValues for notifyCondition %d", ncP->condValueList.size(), ncvIx));
+    for (unsigned int cvIx = 0; cvIx < ncP->condValueList.size(); ++cvIx)
+    {
+      if (ncP->condValueList[cvIx] == attributeName)
+      {
+        LM_M(("KZ: Found matching attribute: '%s'", ncP->condValueList[cvIx].c_str()));
+        return true;
+      }
+    }
+  }
+
+  LM_M(("KZ: No matching attribute for '%s'", attributeName.c_str()));
+  return false;
+}
+
+
+
+/* ****************************************************************************
+*
+* match - 
+*/
+bool Subscription::match
+(
+  const std::string&  _tenant,
+  const std::string&  _servicePath,
+  const std::string&  id,
+  const std::string&  type,
+  const std::string&  attributeName
+)
+{
+  LM_M(("KZ: Matching cache Tenant '%s' to incoming Tenant '%s'", tenant.c_str(), _tenant.c_str()));
+  if (_tenant != tenant)
+  {
+    LM_M(("KZ: Subscription::match: tenant not matching ('%s' != '%s')", tenant.c_str(), _tenant.c_str()));
+    return false;
+  }
+  LM_M(("KZ: Tenant OK"));
+
+  LM_M(("KZ: Matching cache Service-Path '%s' to incoming Service-Path '%s'", servicePath.c_str(), _servicePath.c_str()));
+  if (servicePathMatch(_servicePath) == false)
+  {
+    LM_M(("KZ: Subscription::match: servicePath not matching ('%s' vs '%s')", servicePath.c_str(), _servicePath.c_str()));
+    return false;
+  }
+  LM_M(("KZ: Service-Path OK"));
+
+
+  //
+  // If ONCHANGE and one of the attribute names in the scope vector
+  // of the subscription has the same name as the incoming attribute. there is a match.
+  //
+  LM_M(("KZ: Matching Attributes (incoming: %s)", attributeName.c_str()));
+  if (!attributeMatch(attributeName))
+  {
+    LM_M(("KZ: Subscription::match: attributes not matching (incoming: '%s')", attributeName.c_str()));
+    return false;
+  }
+  LM_M(("KZ: Attributes OK"));
+
+  for (unsigned int ix = 0; ix < entityIdInfos.size(); ++ix)
+  {
+    if (entityIdInfos[ix]->match(id, type))
+    {
+      LM_M(("KZ: Subscription::match: WE HAVE A MATCH !!!"));
+      return true;
+    }
+  }
+
+  LM_M(("KZ: Subscription::match: entityId not matching"));
+  return false;
+}
+
+
+
+/* ****************************************************************************
+*
+* match - 
+*/
+bool Subscription::servicePathMatch
+(
+  const std::string&  _servicePath
+)
+{
+  const char* spath = servicePath.c_str();
+
+  //
+  // No wildcard - exact match
+  //
+  if (spath[strlen(spath) - 1] != '#')
```

Comment:
Let's create the issue. Include the file/fucntion we have detected for the common functionality (that way, in the future we can add other file/fucntions as new comments on that issue).